engine/access/rest/rest_api.go

```diff
@@ -0,0 +1,112 @@
+package rest
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/engine/access/rest/generated"
+	"github.com/onflow/flow-go/engine/access/rpc/backend"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+const MaxBlockIDCount = 50
+
+// RestAPIHandler provides the implementation of each of the REST API
+type APIHandler struct {
+	backend *backend.Backend
+	logger  zerolog.Logger
+}
+
+func NewRestAPIHandler(backend *backend.Backend, logger zerolog.Logger) *APIHandler {
+	return &APIHandler{
+		backend: backend,
+		logger:  logger,
+	}
+}
+
+func (restAPI *APIHandler) BlocksIdGet(w http.ResponseWriter, r *http.Request) {
+	// create a logger for the request
+	errorLogger := restAPI.logger.With().Str("request_url", r.URL.String()).Logger()
+
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+
+	vars := mux.Vars(r)
+	idParam := vars["id"]
+
+	// gorilla mux retains opening and ending square brackets for ids
+	idParam = strings.TrimSuffix(idParam, "]")
+	idParam = strings.TrimPrefix(idParam, "[")
+
+	ids := strings.Split(idParam, ",")
+
+	blocks := make([]*generated.Block, len(ids))
+
+	if len(ids) > MaxBlockIDCount {
+		restAPI.errorResponse(w, http.StatusBadRequest, fmt.Sprintf("at most %d Block IDs can be requested at a time", MaxBlockIDCount), errorLogger)
+		return
+	}
+
+	for i, id := range ids {
+		flowID, err := flow.HexStringToIdentifier(id)
+		if err != nil {
+			restAPI.errorResponse(w, http.StatusBadRequest, fmt.Sprintf("invalid ID %s", id), errorLogger)
+			return
+		}
+
+		flowBlock, err := restAPI.backend.GetBlockByID(r.Context(), flowID)
+		if err != nil {
+			// if error has GRPC code NotFound, the return HTTP NotFound error
+			if status.Code(err) == codes.NotFound {
+				restAPI.errorResponse(w, http.StatusNotFound, fmt.Sprintf("block with ID %s not found", id), errorLogger)
+				return
+			}
+			restAPI.errorResponse(w, http.StatusInternalServerError, fmt.Sprintf("failed to look up block with ID %s", id), errorLogger)
+			return
+		}
+		blocks[i] = toBlock(flowBlock)
+	}
+
+	encodedBlocks, err := json.Marshal(blocks)
+	if err != nil {
+		restAPI.errorResponse(w, http.StatusInternalServerError, err.Error(), errorLogger)
+		return
+	}
+
+	_, err = w.Write(encodedBlocks)
+	if err != nil {
+		restAPI.errorResponse(w, http.StatusInternalServerError, err.Error(), errorLogger)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
```

Comment:
I feel we should extract this to a helper method like `JSONResponse` where we handle serialization and errors and set headers. The reason for that is every handler will reimplement this code. Something like (note code is handwritten as an example, not tested)
```
func JSONResponse(w http.ResponseWriter, payload interface{}) {
	w.Header().Set("Content-Type", "application/json")
	
	response, err := json.Marshal(payload)
	if err != nil {
		restAPI.errorResponse(w, http.StatusInternalServerError, err.Error(), errorLogger)
		return
	}
	
	_, err = w.Write(response)
	if err != nil {
		restAPI.errorResponse(w, http.StatusInternalServerError, err.Error(), errorLogger)
		return
	}
	
	w.WriteHeader(http.StatusOK)
}
```